### ISSUE TITLE
Must call Module::fwd before Module::bwd when calling `backward` for `Encoding` when using the same sets of points multiple times

### ISSUE THREAD

**Issue by kacperkan** [NVlabs/tiny-cuda-nn#204]
Hi everyone,

Thank you for your amazing framework! I'm trying to use just `Encoding` class for my python project. The method works like this:
1. Sample points in the 3D space.
2. Iterate over these points (because of memory constraints).
3. For each iteration, use the `Encoding` encoder.

However, the iteration causes the gradients are not stored, from what I can see. Instead, the computation graph is reset for the latest batch that the backward was used for. It raises the following error:
`RuntimError: Must call Module::fwd before Module::bwd`

My setup:
- RTX 2080 Ti
- tiny-cuda-nn 1.4
- PyTorch 1.8.2
- CUDA 11.1
- Ubuntu 20.04

I'm sharing an MWE below to reproduce the error:
```python
import tinycudann as  tcnn
import torch

encoder = tcnn.Encoding(
    3,
    {
        "otype": "Grid",
        "type": "Hash",
        "n_levels": 16,
        "n_features_per_level": 2,
        "base_resolution": 16,
        "log2_hashmap_size": 20,
        "interpolation": "Linear",
        "dtype": "float32",
    }
).cuda()
sum(encoder(torch.ones((1, 3)).cuda()).sum() for _ in range(2)).backward() # this line should raise the error
```

I will appreciate any help here!

Best regards,
kacperkan

EDIT:
The error happens even with the line:
`sum(encoder(torch.rand((1, 3)).cuda()).sum() for _ in range(2)).backward()`,
so it's enough to call `Encoding` multiple times to break it.

EDIT2: I'm limited by the PyTorch version (other dependencies that rely specifically on 1.8.2), and I'd prefer to stick to tiny-cuda-nn==1.4. Updating tiny-cuda-nn raises the same issue for me as the issue here: https://github.com/NVlabs/tiny-cuda-nn/issues/84

**Comment by Tom94** [comment]
Hi, unfortunately, I don't think the problem can be solved without upgrading to a newer version of tiny-cuda-nn. The old versions simply don't have the capability to keep track of multiple forward passes prior to calling backward (which can be needed in use cases like yours).

This seems to ultimately boil down to a fix for #84 being required

**Comment by Tom94** [comment]
Thanks again for contributing. I think this problem is sorted now, but please feel free to re-open the issue if it isn't.
